pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,1438 @@
+				// This volume was dynamically provisioned for this claim. The
+				// claim got bound elsewhere, and thus this volume is not
+				// needed. Delete it.
+				if err = ctrl.reclaimVolume(volume); err != nil {
+					// Deletion failed, we will fall back into the same condition
+					// in the next call to this method
+					return err
+				}
+				return nil
+			} else {
+				// Volume is bound to a claim, but the claim is bound elsewhere
+				// and it's not dynamically provisioned.
+				if ctrl.hasAnnotation(volume.ObjectMeta, annBoundByController) {
+					// This is part of the normal operation of the controller; the
+					// controller tried to use this volume for a claim but the claim
+					// was fulfilled by another volume. We did this; fix it.
+					glog.V(5).Infof("Synchronizing PersistentVolume[%s]: volume is bound by controller to a claim that is bound to another volume, unbinding", volume.Name)
+					if err = ctrl.unbindVolume(volume); err != nil {
+						return err
+					}
+					return nil
+				} else {
+					// The PV must have been created with this ptr; leave it alone.
+					glog.V(5).Infof("Synchronizing PersistentVolume[%s]: volume is bound by user to a claim that is bound to another volume, waiting for the claim to get unbound", volume.Name)
+					// This just updates the volume phase and clears
+					// volume.Spec.ClaimRef.UID. It leaves the volume pre-bound
+					// to the claim.
+					if err = ctrl.unbindVolume(volume); err != nil {
+						return err
+					}
+					return nil
+				}
+			}
+		}
+	}
+}
+
+// Run starts all of this controller's control loops
+func (ctrl *PersistentVolumeController) Run() {
+	glog.V(5).Infof("Starting PersistentVolumeController")
+	if ctrl.stopChannels == nil {
+		ctrl.stopChannels = make(map[string]chan struct{})
+	}
+
+	if _, exists := ctrl.stopChannels["volumes"]; !exists {
+		ctrl.stopChannels["volumes"] = make(chan struct{})
+		go ctrl.volumeController.Run(ctrl.stopChannels["volumes"])
+	}
+
+	if _, exists := ctrl.stopChannels["claims"]; !exists {
+		ctrl.stopChannels["claims"] = make(chan struct{})
+		go ctrl.claimController.Run(ctrl.stopChannels["claims"])
+	}
+}
+
+// Stop gracefully shuts down this controller
+func (ctrl *PersistentVolumeController) Stop() {
+	glog.V(5).Infof("Stopping PersistentVolumeController")
+	for name, stopChan := range ctrl.stopChannels {
+		close(stopChan)
+		delete(ctrl.stopChannels, name)
+	}
+}
+
+// isFullySynced returns true, if both volume and claim caches are fully loaded
+// after startup.
+// We do not want to process events with not fully loaded caches - e.g. we might
+// recycle/delete PVs that don't have corresponding claim in the cache yet.
+func (ctrl *PersistentVolumeController) isFullySynced() bool {
+	return ctrl.volumeController.HasSynced() && ctrl.claimController.HasSynced()
+}
+
+func (ctrl *PersistentVolumeController) hasAnnotation(obj api.ObjectMeta, ann string) bool {
```

Comment:
this and `setAnnotation` can be stateless methods -- I'm surprised there isn't a utility for this, since this is so general.